include/openssl/digest.h

```diff
@@ -171,6 +171,12 @@ OPENSSL_EXPORT int EVP_DigestUpdate(EVP_MD_CTX *ctx, const void *data,
 // at least this much space.
 #define EVP_MAX_MD_SIZE 64  // SHA-512 is the longest so far.
 
+// EVP_MAX_MD_CHAINING_LENGTH is the largest chaining length supported, in
+// bytes. This constant is only for Merkle-Damgard-based hashed functions
+// like SHA-1, SHA-2, and MD5.
+// This constant is only used internally by HMAC.
+#define EVP_MAX_MD_CHAINING_LENGTH 64  // SHA-512 has the longest chaining length so far
```

Comment:
The chaining length is not necessarily equal to the digest length: for example SHA-224 chaining length is 32 bytes, not 24 bytes.

The maximum chaining length is currently equal to the maximum digest length, but this is not guaranteed in the future. For example, if NIST creates SHA2-768 using a chaining length of 128 bytes, then the maximum digest length would be 96 bytes but the maximum chaining length would be 128 bytes. Another example is if NIST adds a SHA-4 with digest length 128 bytes which does not use a Merkle-Damgard construction. Then, the maximum chaining length would stay at 64 bytes, while the maximum digest length would increase to 128 bytes.

Reply:
Do you mean keeping the name `EVP_MAX_MD_CHAINING_LENGTH`?

Reply:
Sorry, Fabrice, I didn't write what I was thinking. I wanted to suggest `EVP_MAX_MD_STATE_SIZE` instead.

Reply:
On a second thought, `state_size` may be confusing because the state such as `struct sha256_state_st` for example in `sha.h` contains other values. If "chaining_length" is the correct term used in standards or elsewhere, we can keep it. I suggest you explain it here as the internal state of the MD algorithm or something of the sort.

Reply:
It is the term used by openssl/AWS LC there:
https://github.com/aws/aws-lc/blob/4e54dd8363396f257d7a2317c48101e18170e6fb/crypto/fipsmodule/digest/md32_common.h#L78-L80
But it's true I have not really found any other place where it's used.

[Wikipedia](https://en.wikipedia.org/wiki/SHA-2) uses "internal state size".
But the issue with "state size" is that it may be confused with the size of `struct sha256_state_st` which is bigger than just the chaining length.

[FIPS 180-4](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf) uses the term "working variables", but "working variables length" would be very confusing.

So at the end of the day, I think chaining length might be the best option.

Reply:
Thanks, Fabrice. I agree to keeping it as "chaining_length" with the added explanation here similarly to md32_common.h: "this is the internal state size without truncation (e.g. 32 for SHA-224 and SHA-256, 64 for SHA-384 and SHA-512)". Does this sound reasonable?